src/storage-types/src/sources.rs

```diff
@@ -2209,6 +2221,61 @@ mod tests {
         });
     }
 
+    #[mz_ore::test]
+    #[cfg_attr(miri, ignore)]
+    fn backward_compatible_migrate() {
+        fn is_sorted(array: &dyn Array) -> bool {
+            let Ok(cmp) = build_compare(array, array) else {
+                // TODO: arrow v51.0.0 doesn't support comparing structs. When
+                // we migrate to v52+, the `build_compare` function is
+                // deprecated and replaced by `make_comparator`, which does
+                // support structs. At which point, this will work (and we
+                // should switch this early return to an expect, if possible).
+                return false;
+            };
+            (0..array.len())
+                .tuple_windows()
+                .all(|(i, j)| cmp(i, j).is_le())
+        }
+        fn testcase(old: &RelationDesc, new: &RelationDesc, datas: &[SourceData]) {
+            fn data_type(schema: &impl Schema2<SourceData>) -> arrow::datatypes::DataType {
+                use mz_persist_types::columnar::ColumnEncoder;
+                let array = Schema2::encoder(schema).expect("valid schema").finish();
+                Array::data_type(&array).clone()
+            }
+
+            let Some(migration) = backward_compatible(&data_type(old), &data_type(new)) else {
+                return;
+            };
+            let mut encoder = Schema2::<SourceData>::encoder(old).expect("valid schema");
+            for data in datas {
+                encoder.append(data);
+            }
+            let old = encoder.finish();
+            let new = Schema2::<SourceData>::encoder(new)
+                .expect("valid schema")
+                .finish();
+            let old: Arc<dyn Array> = Arc::new(old);
+            let new: Arc<dyn Array> = Arc::new(new);
+            let migrated = migration.migrate(Arc::clone(&old));
+            assert_eq!(migrated.data_type(), new.data_type());
+
+            // Check the sortedness preservation, if we can.
+            if migration.preserves_order() && is_sorted(&old) {
+                assert!(is_sorted(&new))
+            }
+        }
+
+        let strat = (any::<RelationDesc>(), any::<RelationDesc>()).prop_flat_map(|(old, new)| {
+            proptest::collection::vec(arb_source_data_for_relation_desc(&old), 2)
+                .prop_map(move |datas| (old.clone(), new.clone(), datas))
+        });
```

Comment:
This is great! But I've found in the past for similar issues comparing two random `RelationDesc`s often doesn't yield interesting scenarios, a more effective strategy is starting with one `RelationDesc` and then modifying it. e.g. something like:

```
enum SchemaMigrationAction {
    AddColumn { name: String, typ: ColumnType },
    DropColumn { name: String },
}

impl SchemaMigrationAction {
    pub fn apply(&self, desc: &mut RelationDesc) { ... }
}

fn arb_schema_migration_actions(desc: &RelationDesc) -> impl Strategy<Value = Vec<SchemaMigrationAction>> { ... }


// ...

#[mz_ore::test]
#[cfg_attr(miri, ignore)]
fn backward_compatible_migrate() {
    let strat = any::<RelationDesc>().prop_flat_map(|desc| {
            let modifications = arb_schema_migration_actions(&desc);
            (desc, modifications)
        })
        .prop_map(|(desc, mods)| {
            let mut new = desc.clone(); 
            new.apply(mods);
            (desc, new)
        })
        .prop_flat_map(|(old, new)| {
            proptest::collection::vec(arb_source_data_for_relation_desc(&old), 2)
                .prop_map(move |datas| (old.clone(), new.clone(), datas))
        });

    proptest!|((old, new, datas) in strat)| { ... } );
}
```

Reply:
I'm also more than happy to add a test like this afterwards, so don't block on it :)

Reply:
I considered this! The issue with this sort of approach is that it then only generates the modifications that you program into it. And those are exactly the sort of things you probably had in mind when writing the code. IMO these sorts of proptests really shine in situations where they generate a scenario that you would never have thought of.

That said, absolutely worthwhile to add a second proptest along the lines of what you said above. It would also have the benefit of allowing us to verify that things we expect to have a migration do.

Reply:
Totally makes sense! This Friday I'll add a second proptest that's along the lines of what I mentioned

Reply:
FWIW, approaches I've seen for this include:
- "Biased" generators - for example, if I'm writing a field-name generator, I might take an existing type as input and reuse field names from that input type with much higher probability.
- Making the universe of generated values artificially small. (For example, only generating a single non-struct array type, or limiting field names to one of four values.)

The former is more powerful but the latter is way easier, if you've got some sense of the types of bugs you're likely to have.

(Not necessary for this PR; I just think this stuff is interesting.)